src/ddmd/root/filename.d

```diff
@@ -721,3 +722,112 @@ nothrow:
+        import core.stdc.errno: errno, EEXIST;
+
+        const createRet = path.extendedPathThen!(p => CreateDirectoryW(&p[0],
+                                                                       null /*securityAttributes*/));
+        const lastError = GetLastError();
+
+        // Preserve compatibility with mkdir since the calling code expects
+        // errno to be set.
+        if (createRet == 0 && lastError == ERROR_ALREADY_EXISTS)
+            errno(EEXIST);
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
+
+    // Converts a path to one suitable to be passed to Win32 API
+    // functions that can deal with paths longer than 248
+    // characters then calls the supplied function on it.
+    // For more information:
+    // https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx
+    package auto extendedPathThen(alias F)(const(char*) path)
+    {
+        import core.sys.windows.winbase: GetFullPathNameW;
+
+        wchar[1024] wpathBuf;
+        // Since the unicode Win32 APIs use UTF16, we need to convert
+        // from UTF8 to UTF16 first.
+        const wpath = path.toWStringz(wpathBuf);
+
+        // Now wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+        static immutable prefix = `\\?\`w;
+
+        // We don't know how large the absolute file path will be.
+        // We need to tell GetFullPathNameW the size of the buffer we're
+        // passing in so we defensively allocate 4 times the size of
+        // the string we have currently.
+        const absLength = prefix.length + wpath.length * 4;
+        static wchar[1024] absBuf;
+        auto absPath = absLength > absBuf.length ? new wchar[absLength] : absBuf[];
+
+        absPath[0 .. prefix.length] = prefix[];
+
+        const absPathRet = GetFullPathNameW(&wpath[0],
+                                            absPath.length,
+                                            &absPath[prefix.length],
+                                            null /*filePartBuffer*/);
+
+        auto extendedPath = absPathRet > absPath.length ? null : absPath[0 .. absPathRet];
+        return F(extendedPath);
+    }
+
+
+    // Converts an UTF8 null-terminated string to an array of wchar that's null
+    // terminated so it can be passed to Win32 APIs.
+    // buf is passed as a scratch space to store the result. If more memory
+    // is needed then toWstringz allocates on the GC heap instead.
+    private wchar[] toWStringz(const(char*) str, wchar[] buf = []) pure nothrow
+    {
+        import core.stdc.string: strlen;
+
+        const length8 = strlen(str);
+
+        // Conservative estimate that should be enough for any UTF8 string
+        // according to https://tools.ietf.org/html/rfc3629.
+        // The +1 is for the null terminator.
+        const length16 = length8 * 4 + 1;
```

Comment:
I don't get it. Worst case utf-8 to utf-16 is a x2 blowup if all characters are in 0 - 127, right?

Reply:
and that's only in the number of bytes. Given that wchar is double the size of char, the correct line would be `const length16 = length8 + 1`;

Reply:
After thinking about it for a while, you're right about the correct line, but maybe not the explanation. I completely confused length in `wchar` with number of bytes, but the way to reason about is this: _if_ the UTF16 encoding needs 2 code units, then the corresponding UTF8 encoding needs at least 2 bytes, so the worst case scenario is UTF16 has a one-to-one correspondence in the string lengths.

Reply:
There are no code points in utf16 that require more code units than the same code point in utf8. I think my reasoning before is right as well, the only case where you have to use the same number of code units in utf16 is for the first 128